File: external_sort.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Внешняя сортировка на Python
Сортирует большие файлы, которые не помещаются в память
"""

import csv
import os
import tempfile
import shutil
import heapq
import time
import sys


class ExternalSortPython:
    """Класс для внешней сортировки"""

    # Какому ключу какой номер соответствует
    SORT_KEYS = {
        0: ('id', 'int'),
        1: ('full_name', 'str'),
        2: ('department', 'str'),
        3: ('position', 'str'),
        4: ('salary', 'int'),
        5: ('is_fired', 'str'),
        6: ('fire_date', 'str')
    }

    def __init__(self, input_file: str, output_file: str = "sorted.txt", sort_key: int = 0):
        self.input_file = input_file
        self.output_file = output_file
        self.sort_key = sort_key
        self.temp_dir = None
        self.temp_files = []

        # Собираем статистику
        self.stats = {
            'file_size': 0,
            'memory_limit': 0,
            'num_chunks': 0,
            'split_time': 0,
            'merge_time': 0,
            'total_time': 0,
            'records_count': 0
        }

        # Размер файла и лимит памяти (10%)
        self.stats['file_size'] = os.path.getsize(input_file)
        self.stats['memory_limit'] = self.stats['file_size'] // 10

        print(f"\n{'=' * 60}")
        print(f"ВНЕШНЯЯ СОРТИРОВКА (PYTHON)")
        print(f"{'=' * 60}")
        print(f"Входной файл: {input_file}")
        print(f"Выходной файл: {output_file}")
        print(f"Ключ сортировки: {self.SORT_KEYS[sort_key][0]}")
        print(f"Размер файла: {self.stats['file_size'] / (1024 * 1024):.1f} МБ")
        print(f"Лимит памяти (10%): {self.stats['memory_limit'] / (1024 * 1024):.1f} МБ")
        print(f"{'=' * 60}\n")

    def __del__(self):
        """При удалении объекта чистим временные файлы"""
        self.cleanup()

    def cleanup(self):
        """Удаляем временные файлы и папку"""
        if self.temp_files:
            for temp_file in self.temp_files:
                if os.path.exists(temp_file):
                    try:
                        os.remove(temp_file)
                    except:
                        pass

        if self.temp_dir and os.path.exists(self.temp_dir):
            try:
                shutil.rmtree(self.temp_dir)
            except:
                pass

    def get_key_func(self):
        """Возвращает функцию, которая достает значение для сортировки"""
        key_name, key_type = self.SORT_KEYS[self.sort_key]

        def key_func(row):
            value = row[self.sort_key]
            if key_type == 'int':
                return int(value) if value else 0
            elif key_type == 'str':
                return value if value else ''
            return value

        return key_func

    def read_chunks(self):
        """Читаем файл блоками по 10% от размера"""
        chunk = []
        chunk_memory = 0

        with open(self.input_file, 'r', encoding='utf-8') as f:
            reader = csv.reader(f)
            header = next(reader)  # запоминаем заголовок

            for row in reader:
                # Примерно считаем сколько памяти занимает строка
                row_size = sum(len(str(cell)) for cell in row)
                chunk_memory += row_size
                chunk.append(row)

                # Если набрали лимит - отдаем чанк
                if chunk_memory >= self.stats['memory_limit']:
                    yield header, chunk
                    chunk = []
                    chunk_memory = 0

            # Последний чанк
            if chunk:
                yield header, chunk

    def sort_and_write_chunk(self, header, chunk, chunk_idx):
        """Сортируем чанк и сохраняем во временный файл"""
        key_func = self.get_key_func()
        chunk.sort(key=key_func)

        temp_file = os.path.join(self.temp_dir, f"chunk_{chunk_idx:04d}.csv")

        with open(temp_file, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(header)
            writer.writerows(chunk)

        return temp_file

    def split_phase(self):
        """Фаза разбиения: читаем, сортируем, сохраняем"""
        print("ФАЗА 1: РАЗБИЕНИЕ")
        print("-" * 40)

        start_time = time.time()

        # Создаем временную папку
        self.temp_dir = tempfile.mkdtemp(prefix="ext_sort_py_")

        chunk_count = 0
        # Считаем количество записей
        records_count = 0

        for header, chunk in self.read_chunks():
            temp_file = self.sort_and_write_chunk(header, chunk, chunk_count)
            self.temp_files.append(temp_file)
            records_count += len(chunk)
            chunk_count += 1

            if chunk_count % 10 == 0:
                print(f"  Создано чанков: {chunk_count}")

        elapsed = time.time() - start_time
        self.stats['split_time'] = elapsed
        self.stats['num_chunks'] = chunk_count
        self.stats['records_count'] = records_count

        print(f"  Создано {chunk_count} чанков")
        print(f"  Время разбиения: {elapsed:.2f} сек\n")

        return chunk_count

    def merge_phase(self):
        """Фаза слияния: объединяем чанки в один файл через кучу"""
        print("ФАЗА 2: СЛИЯНИЕ")
        print("-" * 40)

        start_time = time.time()

        if not self.temp_files:
            raise ValueError("Нет временных файлов для слияния")

        files = []
        readers = []

        try:
            # Открываем все временные файлы
            for temp_file in self.temp_files:
                f = open(temp_file, 'r', encoding='utf-8')
                reader = csv.reader(f)
                header = next(reader)  # заголовок читаем, он одинаковый во всех
                files.append(f)
                readers.append(reader)

            key_func = self.get_key_func()

            # Создаем кучу из первых записей
            heap = []
            for idx, reader in enumerate(readers):
                try:
                    row = next(reader)
                    heapq.heappush(heap, (key_func(row), idx, row))
                except StopIteration:
                    pass

            # Пишем результат
            with open(self.output_file, 'w', newline='', encoding='utf-8') as out_f:
                writer = csv.writer(out_f)
                writer.writerow(header)

                output_count = 0
                while heap:
                    _, idx, row = heapq.heappop(heap)
                    writer.writerow(row)
                    output_count += 1

                    if output_count % 50000 == 0:
                        print(f"  Записано записей: {output_count:,}")

                    # Берем следующую запись из того же файла
                    try:
                        next_row = next(readers[idx])
                        heapq.heappush(heap, (key_func(next_row), idx, next_row))
                    except StopIteration:
                        pass

            print(f"  Записано {output_count:,} записей")

        finally:
            for f in files:
                f.close()

        elapsed = time.time() - start_time
        self.stats['merge_time'] = elapsed

        print(f"  Время слияния: {elapsed:.2f} сек\n")

    def sort(self):
        """Запускаем сортировку"""
        total_start = time.time()

        try:
            num_chunks = self.split_phase()

            if num_chunks == 0:
                raise ValueError("Нет данных для сортировки")

            self.merge_phase()

        except Exception as e:
            print(f"Ошибка при сортировке: {e}")
            raise
        finally:
            self.cleanup()

        total_elapsed = time.time() - total_start
        self.stats['total_time'] = total_elapsed

        # Выводим итоги
        self.print_statistics()

        return self.stats

    def print_statistics(self):
        """Печатаем статистику"""
        print(f"\n{'=' * 60}")
        print(f"ИТОГОВАЯ СТАТИСТИКА")
        print(f"{'=' * 60}")
        print(f"Всего записей: {self.stats['records_count']:,}")
        print(f"Количество чанков: {self.stats['num_chunks']}")
        print(f"Время разбиения: {self.stats['split_time']:.2f} сек")
        print(f"Время слияния: {self.stats['merge_time']:.2f} сек")
        print(f"Общее время: {self.stats['total_time']:.2f} сек")

        if self.stats['total_time'] < 600:
            print(f"\nВремя сортировки в пределах нормы (менее 10 минут)")
        else:
            print(f"\nВремя сортировки превышает 10 минут!")

        print(f"{'=' * 60}\n")


def main():
    """Запуск из командной строки"""
    if len(sys.argv) < 2:
        print("Использование: python external_sort.py <input_file> [output_file] [sort_key]")
        print("\nПараметры:")
        print("  input_file  - входной CSV файл")
        print("  output_file - выходной файл (по умолчанию: sorted.txt)")
        print("  sort_key    - ключ сортировки:")
        print("                0 - id")
        print("                1 - full_name")
        print("                2 - department")
        print("                3 - position")
        print("                4 - salary")
        print("                5 - is_fired")
        print("                6 - fire_date")
        sys.exit(1)

    input_file = sys.argv[1]
    output_file = sys.argv[2] if len(sys.argv) > 2 else "sorted.txt"
    sort_key = int(sys.argv[3]) if len(sys.argv) > 3 else 0

    if sort_key not in ExternalSortPython.SORT_KEYS:
        print(f"Ошибка: Неверный ключ сортировки {sort_key}")
        print(f"Допустимые значения: {list(ExternalSortPython.SORT_KEYS.keys())}")
        sys.exit(1)

    sorter = ExternalSortPython(input_file, output_file, sort_key)
    sorter.sort()


if __name__ == "__main__":
    main()

File: external_sort.cpp
/*
 * Внешняя сортировка на C++
 * Сортирует большие файлы, разбивая на чанки
 */

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <sstream>
#include <chrono>
#include <filesystem>
#include <queue>
#include <iomanip>

using namespace std;
namespace fs = std::filesystem;

// Структура одного сотрудника
struct Employee {
    int id;
    string full_name;
    string department;
    string position;
    int salary;
    bool is_fired;
    string fire_date;
    string hire_date;
    string original_line;  // сохраняем строку как есть для записи

    Employee(const string& line) : original_line(line) {
        parseFromLine(line);
    }

    // Разбираем строку CSV на поля
    void parseFromLine(const string& line) {
        stringstream ss(line);
        string token;

        getline(ss, token, ',');
        id = stoi(token);

        getline(ss, full_name, ',');
        getline(ss, department, ',');
        getline(ss, position, ',');

        getline(ss, token, ',');
        salary = stoi(token);

        getline(ss, token, ',');
        is_fired = (token == "True");

        getline(ss, fire_date, ',');
        getline(ss, hire_date, ',');
    }

    string toLine() const {
        return original_line;
    }
};

// Сравнители для разных ключей
struct CompareById {
    bool operator()(const Employee& a, const Employee& b) const {
        return a.id < b.id;
    }
};

struct CompareByName {
    bool operator()(const Employee& a, const Employee& b) const {
        return a.full_name < b.full_name;
    }
};

struct CompareByDepartment {
    bool operator()(const Employee& a, const Employee& b) const {
        return a.department < b.department;
    }
};

struct CompareByPosition {
    bool operator()(const Employee& a, const Employee& b) const {
        return a.position < b.position;
    }
};

struct CompareBySalary {
    bool operator()(const Employee& a, const Employee& b) const {
        return a.salary < b.salary;
    }
};

struct CompareByFired {
    bool operator()(const Employee& a, const Employee& b) const {
        return a.is_fired < b.is_fired;
    }
};

struct CompareByFireDate {
    bool operator()(const Employee& a, const Employee& b) const {
        return a.fire_date < b.fire_date;
    }
};

// Класс внешней сортировки
class ExternalSortCPP {
private:
    string input_file;
    string output_file;
    string temp_dir;
    size_t memory_limit;
    int sort_key;
    vector<string> temp_files;

    // Размер файла
    size_t getFileSize(const string& filename) {
        return fs::file_size(filename);
    }

    // Имя для временного файла
    string getTempFilename() {
        static int counter = 0;
        return temp_dir + "/temp_" + to_string(++counter) + ".csv";
    }

    // Сортировка чанка по выбранному ключу
    void sortChunkByKey(vector<Employee>& employees, int key) {
        switch(key) {
            case 0: sort(employees.begin(), employees.end(), CompareById()); break;
            case 1: sort(employees.begin(), employees.end(), CompareByName()); break;
            case 2: sort(employees.begin(), employees.end(), CompareByDepartment()); break;
            case 3: sort(employees.begin(), employees.end(), CompareByPosition()); break;
            case 4: sort(employees.begin(), employees.end(), CompareBySalary()); break;
            case 5: sort(employees.begin(), employees.end(), CompareByFired()); break;
            case 6: sort(employees.begin(), employees.end(), CompareByFireDate()); break;
            default: sort(employees.begin(), employees.end(), CompareById());
        }
    }

    // Запись отсортированного чанка в файл
    void writeChunk(const vector<Employee>& employees, const string& filename, const string& header) {
        ofstream out(filename);
        out << header << "\n";
        for (const auto& emp : employees) {
            out << emp.toLine() << "\n";
        }
        out.close();
    }

public:
    ExternalSortCPP(const string& input, const string& output, int key = 0)
        : input_file(input), output_file(output), sort_key(key) {

        // Создаем временную папку
        string temp_dir_name = "temp_sort_cpp_" + to_string(
            chrono::system_clock::now().time_since_epoch().count());
        temp_dir = temp_dir_name;
        fs::create_directory(temp_dir);

        // Лимит памяти - 10% от файла
        size_t file_size = getFileSize(input_file);
        memory_limit = file_size / 10;

        cout << "\n============================================================" << endl;
        cout << "ВНЕШНЯЯ СОРТИРОВКА (C++)" << endl;
        cout << "============================================================" << endl;
        cout << "Входной файл: " << input_file << endl;
        cout << "Выходной файл: " << output_file << endl;

        const char* key_names[] = {"id", "full_name", "department", "position",
                                   "salary", "is_fired", "fire_date"};
        cout << "Ключ сортировки: " << key_names[key] << endl;
        cout << "Размер файла: " << file_size / (1024*1024) << " МБ" << endl;
        cout << "Лимит памяти (10%): " << memory_limit / (1024*1024) << " МБ" << endl;
        cout << "============================================================\n" << endl;
    }

    ~ExternalSortCPP() {
        cleanup();
    }

    // Удаляем временные файлы
    void cleanup() {
        for (const auto& f : temp_files) {
            if (fs::exists(f)) {
                fs::remove(f);
            }
        }
        if (fs::exists(temp_dir)) {
            fs::remove_all(temp_dir);
        }
    }

    // Фаза разбиения
    pair<size_t, double> splitPhase() {
        cout << "ФАЗА 1: РАЗБИЕНИЕ" << endl;
        cout << "----------------------------------------" << endl;

        auto start = chrono::high_resolution_clock::now();

        ifstream in(input_file);
        if (!in.is_open()) {
            throw runtime_error("Не удалось открыть файл: " + input_file);
        }

        string header;
        getline(in, header);

        vector<Employee> buffer;
        size_t current_memory = 0;
        size_t chunk_count = 0;
        size_t total_records = 0;

        string line;
        while (getline(in, line)) {
            current_memory += line.size();
            buffer.emplace_back(line);
            total_records++;

            // Если набрали лимит памяти
            if (current_memory >= memory_limit) {
                sortChunkByKey(buffer, sort_key);
                string temp_file = getTempFilename();
                writeChunk(buffer, temp_file, header);
                temp_files.push_back(temp_file);

                buffer.clear();
                current_memory = 0;
                chunk_count++;

                if (chunk_count % 10 == 0) {
                    cout << "  Создано чанков: " << chunk_count << endl;
                }
            }
        }

        // Последний чанк
        if (!buffer.empty()) {
            sortChunkByKey(buffer, sort_key);
            string temp_file = getTempFilename();
            writeChunk(buffer, temp_file, header);
            temp_files.push_back(temp_file);
            chunk_count++;
        }

        in.close();

        auto end = chrono::high_resolution_clock::now();
        double elapsed = chrono::duration<double>(end - start).count();

        cout << "  Создано " << chunk_count << " чанков" << endl;
        cout << "  Всего записей: " << total_records << endl;
        cout << "  Время разбиения: " << fixed << setprecision(2) << elapsed << " сек\n" << endl;

        return {total_records, elapsed};
    }

    // Фаза слияния
    double mergePhase() {
        cout << "ФАЗА 2: СЛИЯНИЕ" << endl;
        cout << "----------------------------------------" << endl;

        auto start = chrono::high_resolution_clock::now();

        if (temp_files.empty()) {
            throw runtime_error("Нет временных файлов для слияния");
        }

        // Компаратор для кучи (сравнивает записи по выбранному ключу)
        auto cmp = [this](const pair<string, int>& a, const pair<string, int>& b) {
            Employee ea(a.first), eb(b.first);

            switch(sort_key) {
                case 0: return ea.id > eb.id;
                case 1: return ea.full_name > eb.full_name;
                case 2: return ea.department > eb.department;
                case 3: return ea.position > eb.position;
                case 4: return ea.salary > eb.salary;
                case 5: return ea.is_fired > eb.is_fired;
                case 6: return ea.fire_date > eb.fire_date;
                default: return ea.id > eb.id;
            }
        };

        priority_queue<pair<string, int>, vector<pair<string, int>>, decltype(cmp)> pq(cmp);

        vector<ifstream> file_streams;
        string header;

        // Открываем все временные файлы
        for (size_t i = 0; i < temp_files.size(); i++) {
            file_streams.emplace_back(temp_files[i]);
            if (i == 0) {
                getline(file_streams.back(), header);
            } else {
                string dummy;
                getline(file_streams.back(), dummy);
            }

            string line;
            if (getline(file_streams.back(), line)) {
                pq.push({line, i});
            }
        }

        // Сливаем в выходной файл
        ofstream out(output_file);
        out << header << "\n";

        size_t output_count = 0;
        while (!pq.empty()) {
            auto top = pq.top();
            pq.pop();

            out << top.first << "\n";
            output_count++;

            if (output_count % 50000 == 0) {
                cout << "  Записано записей: " << output_count << endl;
            }

            string next_line;
            if (getline(file_streams[top.second], next_line)) {
                pq.push({next_line, top.second});
            }
        }

        out.close();
        for (auto& fs : file_streams) {
            fs.close();
        }

        auto end = chrono::high_resolution_clock::now();
        double elapsed = chrono::duration<double>(end - start).count();

        cout << "  Записано " << output_count << " записей" << endl;
        cout << "  Время слияния: " << fixed << setprecision(2) << elapsed << " сек\n" << endl;

        return elapsed;
    }

    // Запуск сортировки
    void run() {
        auto total_start = chrono::high_resolution_clock::now();

        size_t total_records = 0;
        double split_time = 0, merge_time = 0;

        try {
            auto split_result = splitPhase();
            total_records = split_result.first;
            split_time = split_result.second;

            if (temp_files.empty()) {
                throw runtime_error("Нет данных для сортировки");
            }

            merge_time = mergePhase();

        } catch (const exception& e) {
            cerr << "Ошибка: " << e.what() << endl;
            throw;
        }

        auto total_end = chrono::high_resolution_clock::now();
        double total_time = chrono::duration<double>(total_end - total_start).count();

        // Итоговая статистика
        cout << "\n============================================================" << endl;
        cout << "ИТОГОВАЯ СТАТИСТИКА" << endl;
        cout << "============================================================" << endl;
        cout << "Всего записей: " << total_records << endl;
        cout << "Количество чанков: " << temp_files.size() << endl;
        cout << "Время разбиения: " << fixed << setprecision(2) << split_time << " сек" << endl;
        cout << "Время слияния: " << fixed << setprecision(2) << merge_time << " сек" << endl;
        cout << "Общее время: " << fixed << setprecision(2) << total_time << " сек" << endl;

        if (total_time < 600) {
            cout << "\nВремя сортировки в пределах нормы (менее 10 минут)" << endl;
        } else {
            cout << "\nВремя сортировки превышает 10 минут!" << endl;
        }
        cout << "============================================================\n" << endl;
    }
};

int main(int argc, char* argv[]) {
    string input_file = "data.csv";
    string output_file = "sorted.txt";
    int sort_key = 0;

    if (argc > 1) input_file = argv[1];
    if (argc > 2) output_file = argv[2];
    if (argc > 3) sort_key = atoi(argv[3]);

    try {
        ExternalSortCPP sorter(input_file, output_file, sort_key);
        sorter.run();
    } catch (const exception& e) {
        cerr << "Ошибка: " << e.what() << endl;
        return 1;
    }

    return 0;
}